TTS/tts/layers/overflow/neural_hmm.py

```diff
@@ -0,0 +1,539 @@
+        memory_rnn_dim: int,
+        outputnet_size: List[int],
+        flat_start_params: dict,
+        std_floor: float,
+        use_grad_checkpointing: bool = True,
+    ):
+        super().__init__()
+
+        self.frame_channels = frame_channels
+        self.ar_order = ar_order
+        self.deterministic_transition = deterministic_transition
+        self.prenet_dim = prenet_dim
+        self.memory_rnn_dim = memory_rnn_dim
+        self.use_grad_checkpointing = use_grad_checkpointing
+
+        self.transition_model = TransitionModel()
+        self.emission_model = EmissionModel()
+
+        assert ar_order > 0, f"AR order must be greater than 0 provided {ar_order}"
+
+        self.ar_order = ar_order
+        self.prenet = Prenet(
+            in_features=frame_channels * ar_order,
+            prenet_type=prenet_type,
+            prenet_dropout=prenet_dropout,
+            dropout_at_inference=prenet_dropout_at_inference,
+            out_features=[self.prenet_dim for _ in range(prenet_n_layers)],
+            bias=False,
+        )
+        self.memory_rnn = nn.LSTMCell(input_size=prenet_dim, hidden_size=memory_rnn_dim)
+        self.output_net = Outputnet(
+            encoder_dim, memory_rnn_dim, frame_channels, outputnet_size, flat_start_params, std_floor
+        )
+        self.register_buffer("go_tokens", torch.zeros(ar_order, 1))
+
+    def forward(self, inputs, inputs_len, mels, mel_lens):
+        r"""HMM forward algorithm for training uses logarithmic version of Rabiner (1989) forward algorithm.
+
+        Args:
+            inputs (torch.FloatTensor): Encoder outputs
+            inputs_len (torch.LongTensor): Encoder output lengths
+            mels (torch.FloatTensor): Mel inputs
+            mel_lens (torch.LongTensor): Length of mel inputs
+
+        Shapes:
+            - inputs: (B, T, D_out_enc)
+            - inputs_len: (B)
+            - mels: (B, D_mel, T_mel)
+            - mel_lens: (B)
+
+        Returns:
+            log_prob (torch.FloatTensor): Log probability of the sequence
+        """
+        # Get dimensions of inputs
+        batch_size, N, _ = inputs.shape
+        T_max = torch.max(mel_lens)
+        mels = mels.permute(0, 2, 1)
+
+        # Intialize forward algorithm
+        log_state_priors = self._initialize_log_state_priors(inputs)
+        log_c, log_alpha_scaled, transition_matrix, means = self._initialize_forward_algorithm_variables(mels, N)
+
+        # Initialize autoregression elements
+        ar_inputs = self._add_go_token(mels)
+        h_memory, c_memory = self._init_lstm_states(batch_size, self.memory_rnn_dim, mels)
+
+        for t in range(T_max):
+
+            # Process Autoregression
+            h_memory, c_memory = self._process_ar_timestep(t, ar_inputs, h_memory, c_memory)
+            # Get mean, std and transition vector from decoder for this timestep
+            # Note: Gradient checkpointing currently doesn't works with multiple gpus inside a loop
```

Comment:
If this is a blocker to use multi-gpu we should explain this in the model docstring and docs too

Reply:
Is this a model specific issue or rooted from torch

Reply:
This is a torch issue gradient checkpointing in a loop is currently not supported in DDP. It works just fine for Multi-GPU if we just turn off the flag by `use_grad_checkpointing=False` but will significantly increase memory usage while training.

Reply:
I have added some information please take a look and see if it needs any more explanation.